Enterprise-CMCS/managed-care-review
# MCR-4276 & MCR-4260: add updatesStateAssignment, fetchMcReviewSettings and CMS can view settings page

## Summary
[MCR-4276](https://jiraent.cms.gov/browse/MCR-4276)

- Rename `State` table `users` join to `User` table to `assignedCMSUsers`
- Add `updateStateAssignment` resolver.
  - Only Admin and DMCO CMS users can call this
- Refactor `updateCMSUser` resolver to `updateDivisionAssignment` resolver. 
   - Originally it had the option to update state assignments, but since the authorized users differ from updating state assignments from division assignments I chose to split it out to `updateStateAssignment`.
   - This also simplifies both actions and reduces the chance of accidentally updating both when only trying to update one.
- Add `fetchMcReviewSettings` resolver
   - Very similar to `fetchEmailSettings`, but this returns state analysts from the database with all of the user data for the analyst.
- Refactor `updateCmsUserProperties` Postgres function to skip updating `stateAssignments` with undefined `stateCodes` and return error if `stateCodes` is an empty array.
- Add `read-write-state-assignments` feature flag.

[MCR-4260](https://jiraent.cms.gov/browse/MCR-4260)
- `Settings.tsx` allows CMS and Helpdesk users to view the Settings page when the feature flag is on.
- Use the `fetchMcReviewSettings` API when the feature flag is on to get analyst and email data
- `CMSUsersTable` only displays the division dropdown for admin users.
- On the State analyst tab, it shows all states with empty assignments, this is intentional so that we start assigning analysts to states without them.
   - In the filter dropdown, I filtered out all empty assignments because it was giving each an option in the dropdown. Instead, I added 1 option that is for empty assignments.
   - Also remember filter by assignments is not working, there is a ticket in the epic for fixing that.
  
[MCR-4197](https://jiraent.cms.gov/browse/MCR-4197)
- Fixed graphql-explorer tool for local deployment

#### Related issues

#### Screenshots

#### Test cases covered

`updateStateAssignment`
- Tests run in iteration with authorized and unauthorized users
   - `'allows update state assignments'`
   - `'errors if state assignments are empty, undefined, or null'`
   - `'returns an error with invalid state codes'`
  
`fetchMcReviewSettings`
- `'returns states with assignments'`
- `'returns email configuration'`
- `'errors when called by state user'`

`Settings.test.tsx`
- Refactor tests to run for every authorized user type.
- `'Settings tests for $userRole with read-write-state-assignments on'`
- `'errors if the target is not a CMS user'`
- `'errors if the userID doesnt exist'`
- `'errors if called by a ${userType}'` - authorized users ran in iteration against test.

`EmailSettingsTable.test.tsx`
- Refactor tests to run for every authorized userType.



## QA guidance

- There is no UI to update state assignments, but you can with the GraphQL explorer in the review app.
   - Make sure CMS users from DMCP and OACT divisions cannot call this API mutation
   - Make sure the state analyst setting page shows the updated assignments.
- Check all types of CMS users can view the settings page
   - Only admin users (Admin, Business Owner, and Helpdesk) can see the dropdown for CMS divisions page
- Check that the GraphQL Explorer tool works locally.
